File: cpp/test/Ice/background/EndpointFactory.cpp
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

#ifndef TEST_API_EXPORTS
#    define TEST_API_EXPORTS
#endif

#include <Ice/EndpointFactoryManager.h>

#include <EndpointFactory.h>
#include <EndpointI.h>

using namespace std;

EndpointFactory::EndpointFactory(const IceInternal::EndpointFactoryPtr& factory) : _factory(factory) {}

int16_t
EndpointFactory::type() const
{
    return (int16_t)(EndpointI::TYPE_BASE + _factory->type());
}

string
EndpointFactory::protocol() const
{
    return "test-" + _factory->protocol();
}

IceInternal::EndpointIPtr
EndpointFactory::create(vector<string>& args, bool oaEndpoint) const
{
    return make_shared<EndpointI>(_factory->create(args, oaEndpoint));
}

IceInternal::EndpointIPtr
EndpointFactory::read(Ice::InputStream* s) const
{
    short type;
    s->read(type);
    assert(type == _factory->type());

    s->startEncapsulation();
    IceInternal::EndpointIPtr endpoint = make_shared<EndpointI>(_factory->read(s));
    s->endEncapsulation();
    return endpoint;
}

void
EndpointFactory::destroy()
{
}

IceInternal::EndpointFactoryPtr
EndpointFactory::clone(const IceInternal::ProtocolInstancePtr&) const
{
    return const_cast<EndpointFactory*>(this)->shared_from_this();
}

File: csharp/src/Ice/Internal/EndpointFactory.cs
// Copyright (c) ZeroC, Inc.

namespace Ice.Internal;

using System.Collections.Generic;

public interface EndpointFactory
{
    void initialize();

    short type();

    string protocol();

    EndpointI create(List<string> args, bool oaEndpoint);

    EndpointI read(Ice.InputStream s);

    void destroy();

    EndpointFactory clone(ProtocolInstance instance);
}

public abstract class EndpointFactoryWithUnderlying : EndpointFactory
{
    protected EndpointFactoryWithUnderlying(ProtocolInstance instance, short type)
    {
        instance_ = instance;
        _type = type;
    }

    public void initialize()
    {
        //
        // Get the endpoint factory for the underlying type and clone it with
        // our protocol instance.
        //
        EndpointFactory factory = instance_.getEndpointFactory(_type);
        if (factory != null)
        {
            _underlying = factory.clone(instance_);
            _underlying.initialize();
        }
    }

    public short type()
    {
        return instance_.type();
    }

    public string protocol()
    {
        return instance_.protocol();
    }

    public EndpointI create(List<string> args, bool oaEndpoint)
    {
        if (_underlying == null)
        {
            return null; // Can't create an endpoint without underlying factory.
        }
        return createWithUnderlying(_underlying.create(args, oaEndpoint), args, oaEndpoint);
    }

    public EndpointI read(Ice.InputStream s)
    {
        if (_underlying == null)
        {
            return null; // Can't create an endpoint without underlying factory.
        }
        return readWithUnderlying(_underlying.read(s), s);
    }

    public void destroy()
    {
        if (_underlying != null)
        {
            _underlying.destroy();
        }
        instance_ = null;
    }

    public EndpointFactory clone(ProtocolInstance instance)
    {
        return cloneWithUnderlying(instance, _type);
    }

    public abstract EndpointFactory cloneWithUnderlying(ProtocolInstance instance, short underlying);

    protected abstract EndpointI createWithUnderlying(EndpointI underlying, List<string> args, bool oaEndpoint);

    protected abstract EndpointI readWithUnderlying(EndpointI underlying, Ice.InputStream s);

    protected ProtocolInstance instance_;

    private readonly short _type;
    private EndpointFactory _underlying;
}
